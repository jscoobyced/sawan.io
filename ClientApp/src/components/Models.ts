import { match } from "react-router-dom";

export interface SelectOption {
    value: number;
    text: string;
}

export interface IdParam {
    id: string;
}
