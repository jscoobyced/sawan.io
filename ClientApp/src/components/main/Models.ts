import { NavigationMenuContent } from "../../services/Models";

export interface HeaderProps {
    navigationMenuContent: NavigationMenuContent;
    signIn: (isSignedIn: boolean) => void;
    isSignedIn: boolean;
}
