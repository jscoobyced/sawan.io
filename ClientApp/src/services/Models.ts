export enum Language {
    English = 0,
    French = 1
}

export interface Link {
    text: string;
    url: string;
    target?: string;
    title?: string;
}

export interface MainContent {
    navigationMenuContent: NavigationMenuContent;
    menuContent: MenuContent;
    footerContent: FooterContent;
}

export interface NavigationMenuContent {
    home: string;
    websiteName: string;
    applications: string;
    cryptoCurrency: string;
    healthMonitor: string;
    about: string;
    information: string;
    resume: string;
}

export interface BlogElement {
    id: number;
    blogDate: Date;
    article: string;
    articleTitle: string;
}

export interface BlogPage {
    articles: BlogElement[];
}

export interface MenuContent {
    links: Link[];
}

export interface FooterContent {
    copyright: string;
    credits: string;
    year: string;
}
