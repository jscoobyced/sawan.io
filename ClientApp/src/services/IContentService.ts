import { BlogElement, BlogPage, Language, MainContent } from './Models';

export interface IContentService {
    getMainContent(language: Language): Promise<MainContent>;
    getDefaultMainContent(): MainContent;
    getBlogPage(maxResult: number): Promise<BlogPage>;
    getDefaultBlogPage(): BlogPage;
    getBlogElement(id: number): Promise<BlogElement>;
}
