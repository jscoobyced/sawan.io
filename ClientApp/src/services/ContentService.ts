import { ApiUtils } from '../utils/ApiUtils';
import { IContentService } from './IContentService';
import { BlogElement, BlogPage, Language, MainContent } from './Models';

export class ContentService implements IContentService {

    public async getMainContent(language: Language): Promise<MainContent> {
        return this.fetchMainContent(language);
    }

    public getDefaultMainContent(): MainContent {
        return this.defaultMainContent();
    }

    public async getBlogPage(yearMonth: string, maxResult: number): Promise<BlogPage> {
        return this.updateBlogPage(yearMonth, maxResult);
    }

    public async getBlogElement(id: number): Promise<BlogElement> {
        return ApiUtils.fetchData<BlogElement>(`/api/Data/blog/${id}`);
    }

    public getDefaultBlogPage(): BlogPage {
        return {
            articles: []
        };
    }

    public async saveBlogElement(blogElement: BlogElement): Promise<boolean> {
        return this.doSaveBlogElement(blogElement);
    }

    protected async doSaveBlogElement(blogElement: BlogElement): Promise<boolean> {
        return ApiUtils.postData('/api/Data/blog/post/', blogElement);
    }

    protected async fetchMainContent(language: Language): Promise<MainContent> {
        return ApiUtils.fetchData<MainContent>(`/api/Data/main/${language}`);
    }

    protected defaultMainContent(): MainContent {
        return {
            menuContent: {
                title: '',
                links: []
            },
            navigationMenuContent: {
                about: '',
                applications: '',
                cryptoCurrency: '',
                healthMonitor: '',
                home: '',
                websiteName: '',
                information: '',
                resume: ''
            },
            footerContent: {
                copyright: '',
                credits: '',
                year: ''
            }
        };
    }

    protected async updateBlogPage(yearMonth: string, maxResult: number): Promise<BlogPage> {
        const empty = {
            articles: []
        };
        if (maxResult <= 0) {
            return empty;
        }
        const data = await ApiUtils.fetchData<BlogElement[]>(`/api/Data/blogpage/${yearMonth}/${maxResult}`);
        if (!data || data.length === 0) {
            return empty;
        }
        return {
            articles: data.slice(0, Math.min(maxResult, data.length))
        };
    }
}
