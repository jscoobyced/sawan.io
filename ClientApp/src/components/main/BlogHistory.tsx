import * as React from 'react';
import { NavLink } from 'react-router-dom';
import { MenuContent } from '../../services/Models';

export interface BlogHistoryProps {
    menu: MenuContent;
}

const BlogHistory: React.FC<BlogHistoryProps> = (props: BlogHistoryProps) => {

    const historyLength = props.menu.links.length;
    const [elementIndex, setStartElement] = React.useState(0);
    const [showHistoryDetails, setShowHistoryDetails] = React.useState(-1);

    const onClick = (key: number) => {
        if (showHistoryDetails === key) {
            setShowHistoryDetails(-1);
        } else {
            setShowHistoryDetails(key);
        }
    };

    const viewEarlier = (event: React.MouseEvent<HTMLElement>) => {
        if (elementIndex < historyLength - 3) {
            setStartElement(elementIndex + 1);
            checkSublistDisplay();
        }
    };

    const viewLater = (event: React.MouseEvent<HTMLElement>) => {
        if (elementIndex > 0) {
            setStartElement(elementIndex - 1);
            checkSublistDisplay();
        }
    };

    const checkSublistDisplay = () => {
        if (showHistoryDetails < elementIndex || showHistoryDetails >= elementIndex + 3) {
            setShowHistoryDetails(-1);
        }
    };

    if (historyLength > 0) {
        const startElement = elementIndex;
        const endElement = Math.min(startElement + 3, historyLength);
        const contentList = props.menu.links.slice(startElement, endElement).map((link, key) => {
            const lkey = key + startElement;
            const contentDetails = link.entries.map((entry, ekey) => (
                <li key={ekey}>
                    <NavLink to={'/blog/view/' + entry.url} >{entry.text}</NavLink>
                </li>
            ));
            return (
                <li key={lkey} >
                    <div className='link' onClick={event => onClick(lkey)} data-test={'link-' + lkey}>{link.name}</div>
                    {showHistoryDetails === lkey ? <ul>{contentDetails}</ul> : <></>}
                </li>);
        });
        if (historyLength > 3 && endElement < historyLength) {
            contentList.push(
                <li key='history-ve-1'>
                    <div onClick={viewEarlier} className='link' title='View earlier' data-test={'link-ve'}>
                        <i className='far fa-arrow-down' />
                    </div>
                </li>);
        }
        if (startElement > 0) {
            contentList.unshift(
                <li key='history-vl-1'>
                    <div onClick={viewLater} className='link' title='View Later' data-test={'link-vl'}>
                        <i className='far fa-arrow-up' />
                    </div>
                </li>);
        }
        return (
            <>
                <h1>{props.menu.title}</h1>
                <ul>{contentList}</ul>
            </>);
    }
    return <></>;
};

export default BlogHistory;
