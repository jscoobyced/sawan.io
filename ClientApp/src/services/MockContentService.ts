import { DateUtil } from '../utils/DateUtils';
import { ContentService } from './ContentService';
import { BlogElement, BlogPage, Language, MainContent } from './Models';

export class MockContentService extends ContentService {

    private readonly data = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do'
        + ' ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco'
        + ' aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit '
        + ' cillum dolore eu fugiat nulla pariatur.<br />Excepteur sint occaecat cupidatat non proident, '
        + ' culpa qui officia deserunt mollit anim id est laborum. Commodo qui imperdiet massa tincidunt.'
        + ' Non enim praesent elementum facilisis leo. Adipiscing diam donec adipiscing tristique. Velit'
        + ' euismod in pellentesque massa placerat. Interdum velit euismod in pellentesque. Phasellus'
        + ' vestibulum lorem sed risus ultricies tristique nulla. Egestas erat imperdiet sed euismod nisi'
        + ' porta lorem mollis aliquam. Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque'
        + ' fermentum. Morbi tristique senectus et netus et malesuada fames.<br />Vitae suscipit tellus'
        + ' mauris a diam maecenas. Sed risus ultricies tristique nulla aliquet. Sagittis eu volutpat odio'
        + ' facilisis mauris sit amet. Tempor orci eu lobortis elementum nibh tellus molestie nunc non.'
        + ' Turpis nunc eget lorem dolor.';
    private readonly displayDate = DateUtil.defaultDate();
    private readonly articles: BlogElement[] = [
        {
            id: '1',
            blogDate: this.displayDate,
            articleTitle: 'Starting again',
            article: 'A new begining this year for me. After 5 years at Agoda it is time for me to move on to'
                + ' new challenges. I am going to do something I haven\'t done for a while: writing a '
                + 'resume that reflects my skills and experience without showing off. Not too much at least '
                + '<i class=\'far fa-smile\'></i>'
                + '<br />So I\'m going to take this opportunity to create it using good CSS3 and HTML5 practices.'
                + ' The first thing I have in mind is to limit the DOM depth. Resumes have simple structures, so'
                + ' we shouldn\'t need a long list of nested DIVs. Second is to chosing a monospaced font so we can'
                + ' keep control of the format when printing. Let\'s throw in some '
                + '<a target=\'_blank\' href=\'https://fontawesome.com/\'>font awesome</a> to have a nice touch on '
                + 'the contact details.<br /> You can see the current version <a href=\'/resume/?full\' target=\'_blank\'>'
                + 'here</a>.'
        },
        {
            id: '2',
            blogDate: this.displayDate,
            articleTitle: 'Grand opening of sawan.io',
            article: 'Grand opening of sawan.io.<br />This website is about my'
                + ' personal training on various web-technologies: TypeScript, webpack, ReactJS, Secure HTTP'
                + '  headers... It includes a simple (and a bit naive) crypto-currency analyser: it can'
                + ' suggest you to sell or buy. Use at your own risk, it is amateur predictions, you have'
                + ' been warned.<br />There will be also an application to track health metrics (weight,'
                + ' % body fat, vascular fat ratio...). Or maybe not, this might be my next project.<br />'
                + ' This project is open-sourced on <a target=\'_blank\''
                + ' href=\'https://github.com/jscoobyced/sawan.io\'>Github</a>. Enjoy and have fun.'
        },
        {
            id: '3',
            blogDate: this.displayDate,
            articleTitle: 'Some other news',
            article: this.data
        }
    ];

    public async getBlogElement(id: string): Promise<BlogElement> {
        return this.updateBlogPage(3).then(blogPage => {
            let article = null as unknown as BlogElement;
            for (const a of blogPage.articles) {
                if (a.id === id) {
                    article = a;
                    break;
                }
            }

            return article;
        });
    }

    protected updateMainContent(language: Language): Promise<MainContent> {
        const content: MainContent = {
            menuContent: {
                links: []
            },
            navigationMenuContent: {
                about: 'About',
                applications: 'Applications',
                cryptoCurrency: 'Crypto-Currency',
                healthMonitor: 'Health Monitor',
                home: 'Home',
                websiteName: 'sawan.io',
                information: 'Information',
                resume: 'Resume'
            },
            footerContent: {
                copyright: 'Copyright',
                credits: 'sawan.io',
                year: '2018 - ' + new Date().getFullYear()
            }
        };
        return Promise.resolve(content);
    }

    protected async doSaveBlogElement(blogElement: BlogElement): Promise<boolean> {
        for (const article of this.articles) {
            if (article.id === blogElement.id) {
                article.article = blogElement.article;
                article.articleTitle = blogElement.articleTitle;
            }
        }
        return Promise.resolve(true);
    }

    protected updateBlogPage = (maxResult: number): Promise<BlogPage> => {
        const blog: BlogPage = {
            articles: this.articles
        };
        return Promise.resolve(blog);
    }
}
