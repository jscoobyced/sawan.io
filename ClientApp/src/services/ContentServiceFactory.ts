import { ContentService } from './ContentService';
import { IContentService } from './IContentService';
import { MockContentService } from './MockContentService';

export class ContentServiceFactory {

    public static GetContentService(): IContentService {
        const mode = process.env.mode as string;
        if (mode === 'production') {
            return new ContentService();
        }

        return new MockContentService();
    }
}
