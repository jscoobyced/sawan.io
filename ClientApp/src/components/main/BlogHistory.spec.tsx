import { shallow } from 'enzyme';
import React from 'react';
import { MenuContent } from '../../services/Models';
import BlogHistory from './BlogHistory';

test('BlogHistory component is unchanged.', () => {
    const entries = [
        {
            text: 'One',
            url: '1'
        },
        {
            text: 'Two',
            url: '2'
        },
        {
            text: 'Three',
            url: '3'
        }
    ];

    const menuContent: MenuContent = {
        title: 'History',
        historyMenus: [
            {
                name: 'April',
                entries
            },
            {
                name: 'March',
                entries
            },
            {
                name: 'February',
                entries
            },
            {
                name: 'January',
                entries
            },
            {
                name: 'December',
                entries
            }
        ]
    };

    const blogHistory = shallow(<BlogHistory menu={menuContent} />);

    expect(blogHistory).toMatchSnapshot();
    blogHistory
        .find('[data-test="link-0"]')
        .simulate('click', { preventDefault: () => { return; } });
    blogHistory
        .find('[data-test="link-0"]')
        .simulate('click', { preventDefault: () => { return; } });
    blogHistory
        .find('[data-test="link-ve"]')
        .simulate('click', { preventDefault: () => { return; } });
    blogHistory
        .find('[data-test="link-vl"]')
        .simulate('click', { preventDefault: () => { return; } });
});

test('BlogHistory component is unchanged.', () => {
    const menuContent: MenuContent = {
        title: 'test',
        historyMenus: []
    };
    const blogHistory = shallow(<BlogHistory menu={menuContent} />);
    expect(blogHistory).toMatchSnapshot();
});
