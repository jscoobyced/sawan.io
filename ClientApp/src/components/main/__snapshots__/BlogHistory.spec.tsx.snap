// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogHistory component is unchanged. 1`] = `
<Fragment>
  <h1>
    History
  </h1>
  <ul>
    <li
      key="0"
    >
      <div
        className="link"
        data-test="link-0"
        onClick={[Function]}
      >
        April
      </div>
    </li>
    <li
      key="1"
    >
      <div
        className="link"
        data-test="link-1"
        onClick={[Function]}
      >
        March
      </div>
    </li>
    <li
      key="2"
    >
      <div
        className="link"
        data-test="link-2"
        onClick={[Function]}
      >
        February
      </div>
    </li>
    <li
      key="history-ve-1"
    >
      <div
        className="link"
        data-test="link-ve"
        onClick={[Function]}
        title="View earlier"
      >
        <i
          className="far fa-arrow-down"
        />
      </div>
    </li>
  </ul>
</Fragment>
`;

exports[`BlogHistory component is unchanged. 2`] = `<Fragment />`;
